src/Decoder.c

```diff
@@ -1950,102 +1864,110 @@ static ZyanStatus ZydisDecodeOperands(ZydisDecoderContext* context,
-                        (instruction->operands[i].mem.base == ZYDIS_REGISTER_SP)  ||
-                        (instruction->operands[i].mem.base == ZYDIS_REGISTER_BP))
-                    {
-                        instruction->operands[i].mem.segment = ZYDIS_REGISTER_SS;
-                    } else
+                    operands[i].mem.segment = ZYDIS_REGISTER_SS;
+                }
+                else
+                    if (!operand->ignore_seg_override &&
+                        instruction->attributes & ZYDIS_ATTRIB_HAS_SEGMENT_DS)
                     {
-                        instruction->operands[i].mem.segment = ZYDIS_REGISTER_DS;
+                        operands[i].mem.segment = ZYDIS_REGISTER_DS;
                     }
-                }
-            }
+                    else
+                        if (!operand->ignore_seg_override &&
+                            instruction->attributes & ZYDIS_ATTRIB_HAS_SEGMENT_ES)
+                        {
+                            operands[i].mem.segment = ZYDIS_REGISTER_ES;
+                        }
+                        else
+                            if (!operand->ignore_seg_override &&
+                                instruction->attributes & ZYDIS_ATTRIB_HAS_SEGMENT_FS)
+                            {
+                                operands[i].mem.segment = ZYDIS_REGISTER_FS;
+                            }
+                            else
+                                if (!operand->ignore_seg_override &&
+                                    instruction->attributes & ZYDIS_ATTRIB_HAS_SEGMENT_GS)
+                                {
+                                    operands[i].mem.segment = ZYDIS_REGISTER_GS;
+                                }
+                                else
+                                {
+                                    if (operands[i].mem.segment == ZYDIS_REGISTER_NONE)
+                                    {
+                                        if ((operands[i].mem.base == ZYDIS_REGISTER_RSP) ||
+                                            (operands[i].mem.base == ZYDIS_REGISTER_RBP) ||
+                                            (operands[i].mem.base == ZYDIS_REGISTER_ESP) ||
+                                            (operands[i].mem.base == ZYDIS_REGISTER_EBP) ||
+                                            (operands[i].mem.base == ZYDIS_REGISTER_SP) ||
+                                            (operands[i].mem.base == ZYDIS_REGISTER_BP))
+                                        {
+                                            operands[i].mem.segment = ZYDIS_REGISTER_SS;
+                                        }
+                                        else
+                                        {
+                                            operands[i].mem.segment = ZYDIS_REGISTER_DS;
+                                        }
+                                    }
+                                }
```

Comment:
Not a fan of the rightward drift here, perhaps just going with inverted conditions and `goto`s here might be preferable? Could also do `do { .. } while (ZYAN_FALSE);` and break out of it, but that's essentially just a less honest goto.